File: leetcode/longest-substring-without-repeating-characters/py/ans_03_01.py
#
# @lc app=leetcode id=3 lang=python3
#
# [3] Longest Substring Without Repeating Characters
#

# @lc code=start
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        allMax = 0
        curStr = s
        nxtBegin = 0
        while len(curStr) != 0:
            curMax, curStr, nxtBegin = self.findCurrentMax(curStr, nxtBegin)
            allMax = max(allMax, curMax)
        return allMax

    def findCurrentMax(self, s: str, begin: int):
        curMax = begin
        while curMax < len(s):
            curCh = s[curMax]
            used = s[:curMax]
            if curCh in used:
                idx = used.index(curCh)
                return curMax, s[idx+1:], curMax - idx
            curMax += 1
        return curMax, "", curMax
# @lc code=end

File: leetcode/longest-substring-without-repeating-characters/cpp/ans_03_01.cpp
/*
 * @lc app=leetcode id=3 lang=cpp
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
#include <unordered_map>
#include <map>
using namespace std;
using goMap = unordered_map<char, int>;
// using goMap = map<char, int>;
class Solution
{
private:
    int i = 0;
    int j = 0;

public:
    int lengthOfLongestSubstring(string s)
    {
        int res = 0;
        while (j < s.length())
        {
            res = max(res, findCurrentMax(s));
        }
        return res;
    }

private:
    int findCurrentMax(const string &s)
    {
        goMap used;
        initUsed(s, used);
        while (j < s.length())
        {
            auto curCh = s[j];
            if (isIn(used, curCh))
            {
                int res = j - i;
                i = used[curCh] + 1;
                return res;
            }
            used[curCh] = j;
            j++;
        }
        return j - i;
    }

    bool isIn(const goMap &used, const char &curCh)
    {
        return used.find(curCh) != used.end();
    }

    void initUsed(const string &s, goMap &used)
    {
        for (auto k = i; k < j; k++)
        {
            used[s[k]] = k;
        }
    }
};
// @lc code=end
